File: 01-in-memory/node/accumulator.js
module.exports = (function () {
    let hits = 0
    let acc = 0

    function alter (v) {
        hits++
        acc += v
    }

    return {
        inc: (v = 1) => alter(v),
        dec: (v = 1) => alter(0 - v),
        get: () => ({ hits, acc }),
    }
})()

File: 01-in-memory/cs/Accumulators/Accumulator.cs
namespace InMemory.Accumulators;

public sealed class Accumulator
{
    private int acc;
    private int hits;
    private static Accumulator? _instance = null;

    private Accumulator()
    {
        this.acc = 0;
        this.hits = 0;
    }

    public static Accumulator GetInstance()
    {
        if (_instance == null)
            _instance = new Accumulator();

        return _instance;
    }

    private void alter(int val)
    {
        this.hits++;
        this.acc += val;
    }

    public void Inc(int val)
    {
        this.alter(val);
    }

    public void Dec(int val)
    {
        this.alter(0 - val);
    }

    public AccumulatorStatus Get()
    {
        return new AccumulatorStatus(this.acc, this.hits);
    }
}
